controllers/image_generator.go

```diff
@@ -0,0 +1,407 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/go-logr/logr"
+	configv1 "github.com/openshift/api/config/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+/*
+The image generator builds and deletes pod VM images for a cloud provider. It uses Kubernetes jobs to do this.
+Make sure
+Ensure that all job YAML files meet the following guidelines:
+1. Job files should be located at config/peerpods/
+2. all file names must follow this format: <cloud-provider>-[CVM|VM]-image-[create|delete]-job.yaml
+3. parallelism: 1
+4. completions: 1 (or none)
+5. backoffLimit: 1
+6. restartPolicy: Never
+7. job and pod names describes cloud-provider and operation
+8. Get configuration values from peer-pods-secret and peer-pods-cm volumes and predefined enviroment variables
+9. make sure create/delete jobs are based on the same configuration sources
+10. utilize precreated podvm binaries conatiner if possible
+11. all jobs must abort on image creation failure
+
+Every image creation job must obey to the following guidelines:
+1. create the podvm image and all other required cloud resouces needed for it
+2. Create a container named result that outputs the image ID (and nothing but the image ID) once the image is ready.
+3. image ID can be shared between containers using emptyDir volume
+
+Every image deletion job should preform the following:
+1. based on the exect same configurtion creation job is using
+2. deletes the image and all created resouces
+*/
+
+const (
+	peerpodsCMName          = "peer-pods-cm"
+	peerpodsCMAWSImageKey   = "PODVM_AMI_ID"
+	peerpodsCMAzureImageKey = "AZURE_IMAGE_ID"
+	fipsCMKey               = "BOOT_FIPS"
+	defaultVMType           = "VM"
+)
+
+type ImageGenerator struct {
+	client    client.Client        // controller-runtime client
+	clientset kubernetes.Interface // k8s clientset
+
+	provider     string
+	CMimageIDKey string
+	fips         bool
+}
+
+var ig *ImageGenerator = nil
+var igLogger logr.Logger = ctrl.Log.WithName("image-generator")
+
+// ImageCreate creates a podvm image for a cloud provider if not present
+func ImageCreate(c client.Client) (bool, ctrl.Result) {
+	if ig == nil { // initialize ImageGenerator if not exsits
+		var err error
+		ig, err = newImageGenerator(c)
```

Comment:
Hmm... my previous question on `newImageGenerator()` not returning an error also applies here, and more generally to any error path returning to the operator... why not propagating ?

Reply:
Since the logic has to control the reconciliation loop actions (result and continuation) i let ImageCreate to return it, i avoided adding also the error because of two reasons:
a. it will require to handle 3 return values in multiple functions which i find a bit tedious
b. AFAIU from @pmores the common practice for this operator is not to return error along with ctrl.Result so is there a point in propagating it?

Reply:
We're not doing much with errors except logging them actually, which your PR already takes care for. Agreed that we don't return errors from Reconcile(). Let's go ahead. We still can revisit later if needed.